include/nbl/builtin/glsl/ext/Scan/scan.glsl

```diff
@@ -0,0 +1,146 @@
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(idx, 0u, nbl_glsl_workgroupExclusiveXor, nbl_glsl_workgroupInclusiveXor)
+}
+
+uint nbl_glsl_ext_Scan_upsweepOr(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(idx, 0u, nbl_glsl_workgroupExclusiveOr, nbl_glsl_workgroupInclusiveOr)
+}
+
+uint nbl_glsl_ext_Scan_upsweepAdd(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(idx, 0u, nbl_glsl_workgroupExclusiveAdd, nbl_glsl_workgroupInclusiveAdd)
+}
+
+uint nbl_glsl_ext_Scan_upsweepMul(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(idx, 1u, nbl_glsl_workgroupExclusiveMul, nbl_glsl_workgroupInclusiveMul)
+}
+
+uint nbl_glsl_ext_Scan_upsweepMin(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(idx, uint(-1), nbl_glsl_workgroupExclusiveMin, nbl_glsl_workgroupInclusiveMin)
+}
+
+uint nbl_glsl_ext_Scan_upsweepMax(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_UPSWEEP(idx, 0u, nbl_glsl_workgroupExclusiveMax, nbl_glsl_workgroupInclusiveMax)
+}
+
+//
+//	Downsweep
+//
+
+shared uint global_offset;
+
+#define NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(IDX, IDENTITY, BIN_OP)\
+if (gl_LocalInvocationIndex == (_NBL_GLSL_WORKGROUP_SIZE_ - 1u))\
+	global_offset = nbl_glsl_ext_Scan_getPaddedData(IDX, IDENTITY, false);\
+barrier();\
+uint data = global_offset;\
+if (gl_LocalInvocationIndex != 0u && (gl_GlobalInvocationID.x < nbl_glsl_ext_Scan_Parameters_t_getElementCountPass()))\
+{\
+	uint prev_idx = STRIDED_IDX(gl_GlobalInvocationID.x - 1u);\
+	data = BIN_OP(data, nbl_glsl_ext_Scan_getPaddedData(prev_idx, IDENTITY, false));\
+}\
+barrier();\
+return data;
+
+uint nbl_glsl_ext_Scan_downsweepAnd(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(idx, uint(-1), nbl_glsl_and)
+}
+
+uint nbl_glsl_ext_Scan_downsweepXor(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(idx, 0u, nbl_glsl_xor)
+}
+
+uint nbl_glsl_ext_Scan_downsweepOr(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(idx, 0u, nbl_glsl_or)
+}
+
+uint nbl_glsl_ext_Scan_downsweepAdd(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(idx, 0u, nbl_glsl_add)
+}
+
+uint nbl_glsl_ext_Scan_downsweepMul(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(idx, 1u, nbl_glsl_mul)
+}
+
+uint nbl_glsl_ext_Scan_downsweepMax(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(idx, uint(-1), max)
+}
+
+uint nbl_glsl_ext_Scan_downsweepMin(in uint idx)
+{
+	NBL_GLSL_EXT_SCAN_DEFINE_DOWNSWEEP(idx, 0u, min)
+}
+
+// uint nbl_glsl_ext_Scan_downsweepAdd(in uint idx)
+// {
+// 	if (gl_LocalInvocationIndex == (_NBL_GLSL_WORKGROUP_SIZE_ - 1u))
+// 		global_offset = nbl_glsl_ext_Scan_getPaddedData(idx, 0u, false);
+// 	barrier();
+// 
+// 	uint data = global_offset;
+// 	if (gl_LocalInvocationIndex != 0u && (gl_GlobalInvocationID.x < nbl_glsl_ext_Scan_Parameters_t_getElementCountPass()))
+// 	{
+// 		uint prev_idx = STRIDED_IDX(gl_GlobalInvocationID.x - 1u);
+// 		data += nbl_glsl_ext_Scan_getPaddedData(prev_idx, 0u, false);
+// 	}
+// 	barrier();
+// 	
+// 	return data;
+// }
+
+#define _NBL_GLSL_EXT_SCAN_INCLUDED_
```

Comment:
for header guards this goes at the top, just after the `ifndef` because you want this to be defined before any includes